File: src/sagacraft/core/adventure_loader.py
"""Helpers for loading adventure JSON files."""

from __future__ import annotations

import json
from typing import Any, Dict


def read_adventure_data(adventure_file: str) -> Dict[str, Any]:
    """Read and parse an adventure JSON file.

    Raises:
        OSError: If the file can't be read.
        json.JSONDecodeError: If the file isn't valid JSON.
        TypeError: If the parsed JSON isn't a mapping.
    """
    with open(adventure_file, "r", encoding="utf-8") as handle:
        data: Any = json.load(handle)

    if not isinstance(data, dict):
        raise TypeError("Adventure JSON must be an object at the top level")

    return data

File: rust/sagacraft_rs/src/pyport/core/adventure_loader.rs
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

/// Read and parse an adventure JSON file.
///
/// # Errors
/// - If the file can't be read
/// - If the file isn't valid JSON
/// - If the parsed JSON isn't an object
pub fn read_adventure_data(adventure_file: &Path) -> Result<HashMap<String, Value>, String> {
    let content = fs::read_to_string(adventure_file)
        .map_err(|e| format!("read file {:?}: {}", adventure_file, e))?;

    let data: Value = serde_json::from_str(&content)
        .map_err(|e| format!("parse json {:?}: {}", adventure_file, e))?;

    match data {
        Value::Object(map) => Ok(map),
        _ => Err("Adventure JSON must be an object at the top level".to_string()),
    }
}